interop/lib/interop/client.ex

```diff
@@ -191,19 +185,31 @@ defmodule Interop.Client do
         payload: payload(271_828)
       )
 
-    {:ok, res_enum, %{headers: new_headers}} =
+    {headers, data, trailers} =
       ch
-      |> Grpc.Testing.TestService.Stub.full_duplex_call(metadata: metadata)
+      |> Grpc.Testing.TestService.Stub.full_duplex_call(metadata: metadata, return_headers: true)
       |> GRPC.Stub.send_request(req, end_stream: true)
       |> GRPC.Stub.recv(return_headers: true)
+      |> process_full_duplex_response()
 
     reply = String.duplicate(<<0>>, 314_159)
 
-    {:ok, %{payload: %{body: ^reply}}} =
-      Stream.take(res_enum, 1) |> Enum.to_list() |> List.first()
+    %{payload: %{body: ^reply}} = data
+
+    validate_headers!(headers, trailers)
+  end
 
+  defp process_full_duplex_response({:ok, res_enum, %{headers: new_headers}}) do
+    {:ok, data} = Stream.take(res_enum, 1) |> Enum.to_list() |> List.first()
     {:trailers, new_trailers} = Stream.take(res_enum, 1) |> Enum.to_list() |> List.first()
-    validate_headers!(new_headers, new_trailers)
+    {new_headers, data, new_trailers}
+  end
+
+  defp process_full_duplex_response({:ok, res_enum}) do
+    {:headers, headers} = Stream.take(res_enum, 1) |> Enum.to_list() |> List.first()
+    {:ok, data} = Stream.take(res_enum, 1) |> Enum.to_list() |> List.first()
+    {:trailers, trailers} = Stream.take(res_enum, 1) |> Enum.to_list() |> List.first()
+    {headers, data, trailers}
```

Comment:
I don't get how this works. It seems to me that you're always obtaining the same value from the stream and matching into 3 different formats.

```elixir
iex(1)> s = Stream.cycle([1, 2, 3, 4])
#Function<63.6935098/2 in Stream.unfold/2>
iex(2)> s |> Stream.take(1) |> Enum.to_list()
[1]
iex(3)> s |> Stream.take(1) |> Enum.to_list()
[1]
iex(4)> s |> Stream.take(3) |> Enum.to_list()
[1, 2, 3]
```

I'd expect something closer to the last line instead.

Reply:
So, I kept the behavior similarly to the gun adapter. A few things to make this easier to understand. 

1. We build the Stream struct using [Stream.unfold/2](https://hexdocs.pm/elixir/1.12/Stream.html#unfold/2)
2. the `unfold` function is built in such a way - for both adapters - that the `acc` is map used to find a connection_stream process
3. and the `next_fun` arg is a function that reads directly from the connection_stream that is producing data. 
4. Every time we execute the `next_fun` we read a chunk of data and remove that from the buffer. That action by itself will generate a side effect that will update the state of the connection_stream by removing the chunk of data we just read. 
5. An easier way to visualize that is. Take the code and the execution bellow as an example

### Code 

```elixir
# I used http terms here to make it easier to see the big picture on how this haves 

defmodule MyRequestProcess do
  use GenServer

  def init(_) do
    {:ok, [1, 2, 3]}
  end

  def handle_call(:get, _from, []) do
    {:reply, nil, []}
  end

  def handle_call(:get, _from, [head | tail]) do
    {:reply, head, tail}
  end
end

{:ok, pid} = GenServer.start_link(MyRequestProcess, [:ok])

ex_stream = Stream.unfold(pid, fn request_pid ->
  case GenServer.call(request_pid, :get) do
    nil -> nil
    value -> {value, request_pid}
  end
end)
```

### Execution

```bash
iex(4)> ex_stream |> Stream.take(1) |> Enum.to_list()
[1]
iex(5)> ex_stream |> Enum.to_list()
[2, 3]
iex(6)> ex_stream |> Enum.to_list()
[]
```
----
Final considerations

To make this behave as you pictured on the above comment, for both adapters I would have to somehow store chunks of data that I read from the connection and keep that in memory. 

Also, that would drastically change the current behavior of the lib. If you want to follow that path, I would recommend think in an new option to make the application behave like that and avoid add a breaking change

Reply:
Right! I didn't realize that the stream was a gen server being wrapped. This clarifies things.

Perhaps a comment in the code highlighting this would be good.

Reply:
Mainly, the genserver makes the variable stateful instead of a common stream that we normally encounter

Reply:
Sure! Maybe also add that in the docs about using the streaming?

Reply:
Sounds good!

Reply:
Solved in: https://github.com/elixir-grpc/grpc/pull/272/commits/fe380234597b700f03560b31245455dff3a367fe